pkg/apis/storage/validation/validation_test.go

```diff
@@ -314,8 +332,46 @@ func TestVolumeAttachmentValidation(t *testing.T) {
 				},
 			},
 		},
+		{
+			// no volume source provided
+			ObjectMeta: metav1.ObjectMeta{Name: "no-vol-source"},
+			Spec: storage.VolumeAttachmentSpec{
+				Attacher: "myattacher",
+				NodeName: "node",
+				Source: storage.VolumeAttachmentSource{
+					PersistentVolumeName: nil,
+					InlineVolumeSource:   nil,
+				},
+			},
+		},
+		{
+			// both persistent and inline vol source
+			ObjectMeta: metav1.ObjectMeta{Name: "vol-source-provided"},
+			Spec: storage.VolumeAttachmentSpec{
+				Attacher: "myattacher",
+				NodeName: "node",
+				Source: storage.VolumeAttachmentSource{
+					PersistentVolumeName: &volumeName,
+					InlineVolumeSource:   &storage.InlineVolumeSource{},
+				},
+			},
+		},
+		{
+			// inline volume source missing CSI
+			ObjectMeta: metav1.ObjectMeta{Name: "vol-source-provided"},
+			Spec: storage.VolumeAttachmentSpec{
+				Attacher: "myattacher",
+				NodeName: "node",
+				Source: storage.VolumeAttachmentSource{
+					InlineVolumeSource: &storage.InlineVolumeSource{},
+				},
+			},
+		},
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		utilfeature.DefaultFeatureGate.Set("CSIInlineVolume=true")
```

Comment:
this changes global state and leaves it changed after the test completes. should pass in feature enablement into the validation method or use the `SetFeatureGateDuringTest` helper that restores original state once the test completes

Reply:
added code to turn/off tests